### PR TITLE
Project 2: Enumerable Methods

### PR DESCRIPTION
**Pull request by jebitok-dev**
In this project we:
- setup project and necessary assets including git workflow and Linters
- rebuild enumerable methods like #each #map #inject #select by
- created and tested #my_each_with_index, #my_all?, #my_any?,  #my_none?,  #my_count,  #my_map and  #my_inject
- used yield, call and proc methods

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  # 1 my_each
+  def my_each
+    to_enum(:my_each) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      yield to_a[i]
+      i += 1
+    end
+  end
```
Comment (javitocor) [inline enumerable.rb]: #my_each returns an Enumerator if no block is given
#my_each when a block is given when self is an array returns the array itself after calling the given block once for each element in self
#my_each when a block is given when self is a range returns the range itself after calling the 
given block once for each element in self

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  # 2 my_each_with_index
+  def my_each_with_index
+    to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      yield(to_a[i], i)
+      i += 1
+    end
+  end
```
Comment (javitocor) [inline enumerable.rb]: #my_each_with_index returns an enumerator if no block is given
#my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
#my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+  # 3 my_select
+  def my_select
+    to_enum(:my_select) unless block_given?
+
+    my_arr = []
+    to_a.my_each { |item| my_arr << item if yield item }
+    my_arr
+  end
```
Comment (javitocor) [inline enumerable.rb]: #my_select returns an enumerator if no block is given

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+      to_a.my_each { |i| return true if i }
+    elsif !param.nil? && (param.is_a? Class)
+      to_a.my_each { |i| return true if [i.class, i.class.superclass].include?(param) }
+    elsif !param.nil? && param.instance_of?(Regexp)
+      to_a.my_each { |i| return true if param.match(i) }
+    else
+      to_a.my_each { |i| return true if i == param }
+    end
+    false
+  end
```
Comment (javitocor) [inline enumerable.rb]: #my_any? when no block or argument is given returns true if at least one of the collection is 
not false or nil
#my_any? when no block or argument is given returns false if at least one of the collection is not true
#my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
#my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
#my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
#my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
#my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
#my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+  # 6 my_none
+  def my_none?(param = nil)
+    if block_given?
+      !my_any?(&Proc.new)
+    else
+      !my_any?(param)
+    end
+  end
```
Comment (javitocor) [inline enumerable.rb]: #my_none? when no block or argument is given returns true only if none of the collection members is true
#my_none? when no block or argument is given returns false only if one of the collection members is true
#my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
#my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
#my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
#my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
#my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
#my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
#my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    to_enum(:my_map) unless block_given? || !proc.nil?
+
+    my_arr = []
+    if proc.nil?
+      to_a.my_each { |i| my_arr << yield(i) }
+    else
+      to_a.my_each { |i| my_arr << proc.call(i) }
+    end
+    my_arr
+  end
```
Comment (javitocor) [inline enumerable.rb]: #my_map returns an Enumerator if no block is given

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

Hello team! 👋 
A great job so far! 👍 
Nice code! 💯 
However, we still need to make some changes in order to get the project approved 💪 
Kindly check my comments below 👇 implement the changes and resubmit your code for a new code review ⏩ 
## Mandatory changes
- Remove the examples from your main file, you can create a new file if you want to add them
- With your Enumerator, it is throwing the following error: LocalJumpError, kindly fix them

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle:

### Hi @jebitok-dev   ! :wave: :wave:

#### You have done well so far :clap: However, there are changes required for you to make in order for this project to be approved. Kindly work on them and submit them for another code review.


**Recommended**

- **#my_any**
- [ ] When no block or argument is given returns true if at least one of the collection is
not false or nil
- [ ] When no block or argument is given returns false if at least one of the collection is not true
- [ ] When a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- [ ] When a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- [ ] When a Regex is passed as an argument returns true if any of the collection matches the Regex
- [ ] When a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [ ] When a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

- **#my_none**
- [ ] When no block or argument is given returns true only if none of the collection members is true
- [ ] When a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] When a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- [ ] When a Regex is passed as an argument returns true if none of the collection matches the Regex
- [ ]  When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
     

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Great job! 👏  Both of you have done well thus far on this project.
Let's fix a few bugs and we will be good to go.

- [ ] `#my_any?` when no block or argument is given returns true if at least one of the collection is not false or nil:  `[nil, false, true, []].my_any? == [nil, false, true, []].any?`
- [ ] `#my_any? ` when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern:  `%w[dog door rod blade].my_any?('cat') == %w[dog door rod blade].any?('cat')`
- [ ] `#my_none?` returns true if the block never returns true for all elements : 
```
true_block = [nil, false, true, []]
array.my_none?(&true_block) == array.none?(&true_block)

```

- [ ] `#my_none?` when no block or argument is given returns true only if none of the collection members is true: `[nil, false, nil, false].my_none?` should return `true`
- [ ] `#my_none?` when no block or argument is given returns false only if one of the collection members is true: `[nil, false, true, []].my_none?` should return `false`
- [ ] `#my_none?` when a Regex is passed as an argument returns true if none of the collection matches the Regex: `%w[dog door rod blade].my_none?(/d/) ` should return `false`
- [ ] `#my_none?` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern `%w[dog door rod blade].my_none?(5) ` should return `true`

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by Temmarie — APPROVED** [review]
## Status: Approved :heavy_check_mark:

### Hi @jebitok-dev ! 👋 

#### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: :tada:

#### You may merge, good luck with your next projects! :rocket:
#### Happy Coding! :+1:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Pull request merged**